### ISSUE TITLE
call "singularity exec" from within active singularity shell

### ISSUE THREAD

**Issue by mike-dubman** [apptainer/singularity#4037]
## Version of Singularity:

3.3.0-rc.3
### Expected behavior

```
$singularity shell test.sif
nested-test$module load tools/singularity
nested=test$singularity exec lolcow_latest.sif id
ERROR  : Failed to set effective UID to 0
$
```
* ```started-suid``` utility has setuid bit set and singularity compiled with setuid support 
* non nested it works
* running as non-priv user

Is there way to make it work? or only option is run ```ssh``` out of container into host and launch from there?

### Steps to reproduce behavior

see above

**Comment by jscook2345** [comment]
There's probably a few security mechanisms blocking this. Has anyone done `singularity` in `singularity` ?

**Comment by jmstover** [comment]
> ERROR  : Failed to set effective UID to 0

This is expected as you are running it. When a container is started we use `PR_NO_NEW_PRIVS`. This means that as a normal user, the kernel blocks the privilege escalation attempt. 

To accomplish this, the CAP_SETUID capability[1] should be all you need, or to start the "host" container as root with `--keep-privs`. Starting as root though means more work starting the containers as different users except root.

[1] - Warning... setting CAP_SETUID is effectively giving a user root on the host system with a small amount of work. It should only be used where you trust all the users, and possibly have other controls in place.

**Comment by mike-dubman** [comment]
thanks !
so, following should do the trick for *non-root* user? (will try it tomorrow)
```
% singularity shell --add-caps CAP_SETUID test.sif
%% singularity shell cow.sif
```

?

**Comment by mike-dubman** [comment]
no luck, tried as non-root user. please advise.

```
% singularity shell -add-caps CAP_SETUID test.sif
WARNING: not authorized to add capability: CAP_SETUID
```

**Comment by jmstover** [comment]
With capabilities, you first need to give yourself the ability to use it. So you would run something like:

```
$ sudo singularity capability add --user [user name] CAP_SETUID
```

Once you've given yourself that capability as root, then you can use the `--add-caps CAP_SETUID` option.

See: https://sylabs.io/guides/3.3/user-guide/cli/singularity_capability_add.html

**Comment by DrDaveD** [comment]
I have successfully run unprivileged singularity within singularity on a new enough kernel that supports unprivileged namespaces, in particular RHEL7.6.  No capabilities are needed to be added.  The one limitation I found is that --pid can only be used on one of them.

**Comment by mike-dubman** [comment]
@DrDaveD , @jmstover - thanks for info and help!

@DrDaveD - could you please share setup details and command lines you used to achieve it?

**Comment by DrDaveD** [comment]
@mike-dubman For which operating system & kernel version?

**Comment by mike-dubman** [comment]
> @mike-dubman For which operating system & kernel version?

yes please, also command lines used would be great. will try to reproduce it on my setup.
thanks again!

**Comment by DrDaveD** [comment]
I was asking you what your operating system & kernel version is.  The answer is different depending on those things.

**Comment by mike-dubman** [comment]
> I was asking you what your operating system & kernel version is. The answer is different depending on those things.

ohh.. it is rhel7.4 with vanilla kernel.

**Comment by DrDaveD** [comment]
Then it depends on whether or not you have recently updated the kernel, because the kernel line is shared with all 7.x releases and you need one that was released with 7.6 or later (3.10.0-957 or later).    Then to enable unprivileged user namespaces see [these instructions](https://opensciencegrid.org/docs/worker-node/install-singularity/#enabling-unprivileged-singularity).    After that you should be able to run singularity unprivileged, with a "sandbox" image, not with .sif images, and if you have access to singularity inside your container, you should be able to run it unprivileged.
